src/sqlfluff/core/rules/std/L015.py

```diff
@@ -30,10 +30,42 @@ def _eval(self, segment, raw_stack, **kwargs):
 
         Look for DISTINCT keyword immediately followed by open parenthesis.
         """
-        # We only trigger on start_bracket (open parenthesis)
-        if segment.name == "start_bracket":
-            filt_raw_stack = self.filter_meta(raw_stack)
-            if len(filt_raw_stack) > 0 and filt_raw_stack[-1].name == "DISTINCT":
-                # If we find DISTINCT followed by open_bracket, then bad.
-                return LintResult(anchor=segment)
-        return LintResult()
+        # We only trigger on expressions that begin with start_bracket (open
+        # parenthesis)
+        if segment.type == "expression":
```

Comment:
Would it make sense to first look for select segments and then proceed as-is from there? Otherwise we're going to be digging into every expression.

Reply:
I thought about this a bit, and TBH, I'm unsure. I feel like I don't know SQL extremely well, and there seem to be lots of places you can put things that I didn't know about. I'm worried that limiting it to `SELECT` may cause it to miss other possible scenarios.

Reply:
I have another idea which should help -- the rule begins with two checks:
* `expression`
* `DISTINCT`

I'm going to reverse the order of these two (i.e. checking for outer `DISTINCT` before checking for inner `expression`). I think that'll be much more efficient (`DISTINCT` is presumably much less common than expressions, also expressions can probably be nested but `DISTINCT` typically isn't).